core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java

```diff
@@ -230,13 +233,37 @@ public void clearLockedKeys() {
       lockedKeys = null;
    }
 
-   private boolean hasLockOrIsLockBackup(Object key) {
-      //stopgap fix for ISPN-2728. The real fix would be to synchronize this with the intrinsic lock.
+   @Override
+   public boolean containsLockOrBackupLock(Object key) {
       Set<Object> lockedKeysCopy = lockedKeys;
       Set<Object> backupKeyLocksCopy = backupKeyLocks;
       return (lockedKeysCopy != null && lockedKeysCopy.contains(key)) || (backupKeyLocksCopy != null && backupKeyLocksCopy.contains(key));
    }
 
+   @Override
+   public boolean containsAnyLockOrBackupLock(Stream<Object> keys) {
+      return findFirstLockedOrBackupLocked(keys) != null;
+   }
+
+   @Override
+   public Object findFirstLockedOrBackupLocked(Stream<Object> keys) {
+      Set<Object> lockedKeysCopy = lockedKeys;
+      Set<Object> backupKeyLocksCopy = backupKeyLocks;
+      if (lockedKeysCopy != null && backupKeyLocksCopy != null) {
+         return keys.filter(key -> lockedKeysCopy.contains(key) || backupKeyLocksCopy.contains(key)).findAny().orElse(null);
+      } else if (lockedKeysCopy != null) {
+         return keys.filter(lockedKeysCopy::contains).findAny().orElse(null);
+      } else if (backupKeyLocksCopy != null) {
+         return keys.filter(backupKeyLocksCopy::contains).findAny().orElse(null);
+      }
+      return null;
+   }
```

Comment:
This stack trace talks about it a bit more [1].  The problem is that in some cases it is faster and others it may not be.  But in my findings most cases the performance difference is negligible both ways.  The reason I even mentioned Streams in the first place to Pedro was as a way to prevent additional entry iterations which imo is a much bigger performance issue.

[1] http://stackoverflow.com/questions/22658322/java-8-performance-of-streams-vs-collections#comment36791431_22670380